README.md
# Overview

[![Version](http://img.shields.io/pypi/v/cfg4py?color=brightgreen)](https://pypi.python.org/pypi/cfg4py)
[![CI Status](https://github.com/zillionare/cfg4py/actions/workflows/release.yml/badge.svg)](https://github.com/zillionare/cfg4py)
[![Code Coverage](https://img.shields.io/codecov/c/github/zillionare/cfg4py)](https://app.codecov.io/gh/zillionare/cfg4py)
[![Downloads](https://pepy.tech/badge/cfg4py)](https://pepy.tech/project/cfg4py)
[![License](https://img.shields.io/badge/License-MIT.svg)](https://opensource.org/licenses/MIT)
[![Style](https://img.shields.io/badge/code%20style-black-000000.svg)](https://github.com/psf/black)

* Free software: MIT license
* Documentation: https://zillionare.github.io/cfg4py


A python config module that:

1. Adaptive deployment (default, dev, test, production) support
2. Cascading configuration (central vs local) support
3. Auto-complete
4. Templates (logging, database, cache, message queue,...)
5. Environment variables macro support
6. Enable logging in one line
7. Built on top of yaml

## Features

It's common to see that you have different settings for development machine, test machine and production site. They share many common settings, but a few of them has to be different.

For example, developers should connect to local database server when performing unittest, and tester should connect to their own database server. All these servers should be deployed separately and no data should be messed up.

Cfg4Py has perfect solution supporting for this: adaptive deployment environment support.

### Adaptive Deployment Environment Support

In any serious projects, your application may run at both development, testing and production site. Except for effort of copying similar settings here and there, sometimes we'll mess up with development environment and production site. Once this happen, it could result in very serious consequence.

To solve this, Cfg4Py developed a mechanism, that you provide different sets for configurations: dev for development machine, test for testing environment and production for production site, and all common settings are put into a file called `defaults`.

cfg4py module knows which environment it's running on by looking up environment variable __cfg4py_server_role__. It should be one of `DEV`, `TEST` and `PRODUCTION`. If nothing found, it means setup is not finished, and Cfg4Py will refuse to work. If the environment is set, then Cfg4Py will read settings from defaults set, then apply update from either of `DEV`, `TEST` and `PRODUCTION` set, according to the environment the application is running on.

!!! important

    Since 0.9.0, cfg4py can still work if __cfg4py_server_role__ is not set, when it work at non-strict mode.

### Cascading design


Assuming you have a bunch of severs for load-balance, which usually share same configurations. So you'd like put the configurations on a central repository, which could be a redis server or a relational database. Once you update configuration settings at central repository, you update configurations for all servers. But somehow for troubleshooting or maintenance purpose, you'd like some machines could have its own settings at a particular moment.

This is how Cfg4Py solves the problem:

1. Configure your application general settings at remote service, then implement a `RemoteConfigFetcher` (Cfg4Py has already implemented one, that read settings from redis), which pull configuration from remote serivce periodically.
2. Change the settings on local machine, after the period you've set, these changes are popluated to all machines.

### Auto-complete

[auto-complete](http://images.jieyu.ai/images/projects/cfg4py/auto-complete.gif)


With other python config module, you have to remember all the configuration keys, and refer to each settings by something like cfg["services"]["redis"]["host"] and etc. Keys are hard to rememb, prone to typo, and way too much tedious.

When cfg4py load raw settigns from yaml file, it'll compile all the settings into a Python class, then Cfg4Py let you access your settings by attributes. Compares the two ways to access configure item:

```python

        cfg["services"]["redis"]["host"]
```
vs:

```python

        cfg.services.redis.host
```

Apparently the latter is the better.

And, if you trigger a build against your configurations, it'll generate a python class file. After you import this file (named 'schema.py') into your project, then you can enjoy code auto-complete!

### Templates

It's hard to remember how to configure log, database, cache and etc, so cfg4py provide templates.

Just run cfg4py scaffold, follow the tips then you're done.

[scaffold](http://images.jieyu.ai/images/projects/cfg4py/scaffold.png)


### Environment variables macro

The best way to keep secret, is never share them. If you put account/password files, and these files may be leak to the public. For example, push to github by accident.

With cfg4py, you can set these secret as environment variables, then use marco in config files. For example, if you have the following in defaults.yaml (any other files will do too):

```text

        postgres:
                dsn: postgres://${postgres_account}:${postgres_password}@localhost
```

then cfg4py will lookup postgres_account, postgres_password from environment variables and make replacement.


### Enable logging with one line

with one line, you can enable file-rotating logging:

```python

    cfg.enable_logging(level, filename=None)
```

### Apply configuration change on-the-fly

Cfg4Py provides mechanism to automatically apply configuration changes without restart your application. For local files configuration change, it may take effect immediately. For remote config change, it take effect up to `refresh_interval` settings.

### On top of yaml

The raw config format is backed by yaml, with macro enhancement. YAML is the best for configurations.



### Credits


This package was created [ppw](https://zillionare.github.io/python-project-wizard)


HISTORY.md

# History

!!! Info
    `(#{number})` means an issue of this project. You may check details of the issue by visiting https://github.com/zillionare/cfg4py/issues/_{number}_

## 0.9.4 (2022-08-25)
* lock apscheduler version to 3.9.1. APScheduler 4.0 has been refactored thus cause importing error.
## 0.9.3 (2022-06-03)

* on apple m1, it's not able to watch file changes, and cause cfg4py fail. This revision will disable hot-reload in such scenario and user can still use all other functions of cfg4py.
* remove support for python 3.6 since it's out of service, and opt 3.10, 3.11 in
* (#4) log settings are now available by `cfg.logging`. 

## 0.9.2 (2021-12-17)

* (#1) hot-reload will now only react to configuration files's change.

## 0.9.0 (2020-12-03)

* add strict mode: default is non-strict mode, which allows you run cfg4py without set environment variable __cfg4py_server_role__

    this is a **break** change. If you've used cfg4py in your project and it worked well, after upgrade to 0.9.0, you have to modify your init code as this:

```python

    cfg4py.init('path_to_config_dir', strict = True)
```
    see more in usage and FAQ document

## 0.8.0 (2020-11-22)

* rename cfg4py_auto_gen.py to schema.py

## 0.7.0 (2020-10-31)

* support environment macro

## 0.5.0 (2020-04-30)

* add command hint, set_server_role
* export envar
* add pip source, conda source

## 0.1.0 (2020-04-06)

* First release on PyPI.










cfg4py/cli.py
import logging
import os
import sys
from typing import Optional

import fire
from ruamel.yaml import YAML

from cfg4py import enable_logging, envar, init

enable_logging()


class Command:
    def __init__(self):
        self.resource_path = os.path.normpath(
            os.path.join(os.path.dirname(__file__), "resources/")
        )
        self.yaml = YAML(typ="safe")  # default, if not specfied, is 'rt' (round-trip)
        self.yaml.default_flow_style = False

        with open(
            os.path.join(self.resource_path, "template.yaml"), "r", encoding="utf-8"
        ) as f:
            self.templates = self.yaml.load(f)

        self.transformed = self._transform()

    def build(self, config_dir: str):
        """Compile configuration files into python script, which is used by IDE's
        auto-complete function

        Args:
            config_dir: The folder where your configuration files located

        Returns:

        """
        if not os.path.exists(config_dir):
            print(f"path {config_dir} not exists")
            sys.exit(-1)

        count = 0
        for f in os.listdir(config_dir):
            if (
                f.startswith("default")
                or f.startswith("dev")
                or f.startswith("test")
                or f.startswith("production")
            ):
                print(f"found {f}")
                count += 1

        if count > 0:
            print(f"{count} files found in total")
        else:
            print("the folder contains no valid configuration files")
            sys.exit(-1)

        try:
            init(config_dir)
            sys.path.insert(0, config_dir)
            from schema import Config  # type: ignore # noqa

            output_file = f"{os.path.join(config_dir, 'schema')}"
            msg = f"Config file is built with success and saved at {output_file}"
            print(msg)
        except Exception as e:  # pragma: no cover
            logging.exception(e)
            print("Config file built failure.")

    def _choose_dest_dir(self, dst):
        if dst is None:
            dst = input("Where should I save configuration files?\n")

        if os.path.exists(dst):
            for f in os.listdir(dst):
                msg = f"The folder already contains {f}, please choose clean one."

                if f in ["defaults.yaml", "dev.yaml", "test.yaml", "production.yaml"]:
                    print(msg)
                    return None
            return dst
        else:
            create = input("Path not exists, create('Q' to exit)? [Y/n]")
            if create.upper() == "Y":
                os.makedirs(dst, exist_ok=True)
                return dst
            elif create.upper() == "Q":
                sys.exit(-1)
            else:
                return None

    def scaffold(self, dst: Optional[str]):
        """Creates initial configuration files based on our choices.
        Args:
            dst:

        Returns:

        """
        print("Creating a configuration boilerplate:")
        dst = self._choose_dest_dir(dst)
        while dst is None:
            dst = self._choose_dest_dir(dst)

        yaml = YAML(typ="safe")  # default, if not specfied, is 'rt' (round-trip)
        with open(
            os.path.join(self.resource_path, "template.yaml"), "r", encoding="utf-8"
        ) as f:
            templates = yaml.load(f)

        print("Which flavors do you want?")
        print("-" * 20)
        prompt = """
        0  - console + rotating file logging
        10 - redis/redis-py (gh://andymccurdy/redis-py)
        11 - redis/aioredis (gh://aio-libs/aioredis)
        20 - mysql/PyMySQL (gh://PyMySQL/PyMySQL)
        30 - postgres/asyncpg (gh://MagicStack/asyncpg)
        31 - postgres/psycopg2 (gh://psycopg/psycopg2)
        40 - mq/pika (gh://pika/pika)
        50 - mongodb/pymongo (gh://mongodb/mongo-python-driver)
        """
        print(prompt)
        chooses = input(
            "Please choose flavors by index, separated each by a comma(,):\n"
        )
        flavors = {}
        mapping = {
            "0": "logging",
            "1": "redis",
            "2": "mysql",
            "3": "postgres",
            "4": "mq",
            "5": "mongodb",
        }
        for index in chooses.strip(" ").split(","):
            if index == "0":
                flavors["logging"] = templates["logging"]
                continue

            try:
                major = mapping[index[0]]
                minor = int(index[1])
                flavors[major] = list(templates[major][minor].values())[0]
            except (ValueError, KeyError):
                print(f"Wrong index {index}, skipped.")
                continue

        with open(os.path.join(dst, "defaults.yaml"), "w", encoding="utf-8") as f:
            f.writelines(
                "#auto generated by Cfg4Py: https://github.com/jieyu-tech/cfg4py\n"
            )
            yaml.dump(flavors, f)

        print(f"Cfg4Py has generated the following files under {dst}:")
        print("defaults.yaml")
        for name in ["dev.yaml", "test.yaml", "production.yaml"]:
            with open(os.path.join(dst, name), "w", encoding="utf8") as f:
                f.writelines(
                    "#auto generated by Cfg4Py: https://github.com/jieyu-tech/cfg4py\n"
                )
                print(name)

        with open(os.path.join(dst, "defaults.yaml"), "r", encoding="utf-8") as f:
            print("Content in defaults.yaml")
            for line in f.readlines():
                print(line.replace("\n", ""))

    def _show_supported_config(self):
        print("Support the following configurations:")
        for key in self.templates.keys():
            item = self.templates.get(key)
            if isinstance(item, dict):
                print(f"  {key}")
            elif isinstance(item, list):
                sub_keys = []
                for sub_item in item:
                    sub_keys.append(f"{key}/{list(sub_item.keys())[0]}")
                print(f"  {key}: {', '.join(sub_keys)}")

    def _transform(self):
        transformed = {}
        for key in self.templates:
            if isinstance(self.templates[key], dict):
                transformed[key] = self.templates[key]
            elif isinstance(self.templates[key], list):
                for item in self.templates[key]:
                    item_key = list(item.keys())[0]
                    transformed[f"{key}/{item_key}"] = item

        return transformed

    def hint(self, what: str = None, usage: bool = False):
        """show a cheat sheet for configurations.
        for example:
            cfg4py hint mysql
        this will print how to configure PyMySQL
        :param what
        :param usage
        """

        if what is None or (
            (what not in self.templates) and what not in self.transformed
        ):
            return self._show_supported_config()

        usage_key = f"{what}_usage"

        if usage_key in self.templates and usage:
            print("Usage:", self.templates.get(usage_key))

        if what in self.templates:
            self.yaml.dump(self.templates[what], sys.stdout)

        if usage_key in self.transformed and usage:
            print("Usage:", self.transformed.get(usage_key))

        if what in self.transformed:
            self.yaml.dump(self.transformed[what], sys.stdout)

    def set_server_role(self):
        print("please add the following line into your .bashrc:\n")
        print(f"export {envar}=DEV\n")
        msg = "You need to change DEV to TEST | PRODUCTION according to its actual role\
         accordingly"
        print(msg)

    def version(self):
        from cfg4py import __version__

        print(
            "Easy config module support code complete, cascading design and apative deployment"
        )
        print(f"version: {__version__}")


def main():
    cmd = Command()  # pragma: no cover
    fire.Fire(
        {  # pragma: no cover
            "build": cmd.build,
            "scaffold": cmd.scaffold,
            "hint": cmd.hint,
            "set_server_role": cmd.set_server_role,
            "--version": cmd.version,
            "-v": cmd.version(),
        }
    )


if __name__ == "__main__":
    main()
